README.md
CakeDC Recaptcha Plugin
=======================

The **Recaptcha** plugin for CakePHP provides spam protection in an easy use helper.

Requirements
------------

* CakePHP 2.5+
* PHP 5.2.8+

Documentation
-------------

For documentation, as well as tutorials, see the [Docs](Docs/Home.md) directory of this repository.

Support
-------

For bugs and feature requests, please use the [issues](https://github.com/CakeDC/recaptcha/issues) section of this repository.

Commercial support is also available, [contact us](http://cakedc.com/contact) for more information.

Contributing
------------

This repository follows the [CakeDC Plugin Standard](http://cakedc.com/plugin-standard). If you'd like to contribute new features, enhancements or bug fixes to the plugin, please read our [Contribution Guidelines](http://cakedc.com/contribution-guidelines) for detailed instructions.

License
-------

Copyright 2007-2014 Cake Development Corporation (CakeDC). All rights reserved.

Licensed under the [MIT](http://www.opensource.org/licenses/mit-license.php) License. Redistributions of the source code included in this repository must retain the copyright notice found in each file.

SecrurityComponent in Cakephp3.x
-------

You will need to add following code to your application in every Template you use recaptcha, if you use Secrurity Component. SecrurityComponent is designed to stop hidden fields to change, which is the main function of Recaptcha.

<?= $this->Form->unlockField('recaptcha-token') ?>
<?= $this->Form->unlockField('g-recaptcha-response') ?>


CHANGELOG.md
Changelog
=========

Release 2.2.4
-------------

https://github.com/CakeDC/recaptcha/tree/2.2.4

 * [f8459df](https://github.com/CakeDC/recaptcha/commit/f8459df) Formatting some code in Setup.md and updating semver
 * [c2c88a8](https://github.com/CakeDC/recaptcha/commit/c2c88a8) Fixing the installation instructions

Release 2.2.3
-------------

https://github.com/CakeDC/recaptcha/tree/2.2.3

 * [ad72132](https://github.com/CakeDC/recaptcha/commit/ad72132) Adding semver
 * [c968276](https://github.com/CakeDC/recaptcha/commit/c968276) Naming a few files uppercased to refelect the plugin standard.
 * [8f5d6c0](https://github.com/CakeDC/recaptcha/commit/8f5d6c0) Fixing the headline of the readme.md
 * [74047b7](https://github.com/CakeDC/recaptcha/commit/74047b7) Working on updating the plugin to the new plugin standard
 * [741cc11](https://github.com/CakeDC/recaptcha/commit/741cc11) Use strict comparison
 * [e3344bd](https://github.com/CakeDC/recaptcha/commit/e3344bd) Cs fixes

Test/Case/AllRecaptchaTest.php
<?php
/**
 * Copyright 2010 - 2014, Cake Development Corporation (http://cakedc.com)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright Copyright 2010 - 2014, Cake Development Corporation (http://cakedc.com)
 * @license MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

class AllRecaptchaTest extends PHPUnit_Framework_TestSuite {

/**
 * Suite define the tests for this suite
 *
 * @return void
 */
	public static function suite() {
		$Suite = new CakeTestSuite('All Recaptcha Plugin tests');
		$basePath = CakePlugin::path('Recaptcha') . DS . 'Test' . DS . 'Case' . DS;
		$Suite->addTestDirectory($basePath . DS . 'View' . DS . 'Helper');
		$Suite->addTestDirectory($basePath . DS . 'Model' . DS . 'Behavior');
		$Suite->addTestDirectory($basePath . DS . 'Controller' . DS . 'Component');
		return $Suite;
	}

}

Test/Fixture/ArticleFixture.php
<?php
/**
 * Copyright 2009-2014, Cake Development Corporation (http://cakedc.com)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright Copyright 2009-2014, Cake Development Corporation (http://cakedc.com)
 * @license MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

/**
 * Article Fixture
 *
 * @package recaptcha
 * @subpackage recaptcha.tests.fixtures
 */
class ArticleFixture extends CakeTestFixture {

/**
 * name property
 *
 * @var string
 */
	public $name = 'Article';

/**
 * fields property
 *
 * @var array
 */
	public $fields = array(
		'id' => array('type' => 'integer', 'key' => 'primary'),
		'title' => array('type' => 'string', 'null' => false),
		'comments' => array('type' => 'integer', 'null' => false, 'default' => '0'));

/**
 * records property
 *
 * @var array
 */
	public $records = array(
		array('title' => 'First Article', 'comments' => 2),
		array('title' => 'Second Article', 'comments' => 0));
}


tests/cases/helpers/recaptcha.test.php
<?php
/**
 * Copyright 2009-2010, Cake Development Corporation (http://cakedc.com)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright Copyright 2009-2010, Cake Development Corporation (http://cakedc.com)
 * @license MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

App::import('Core', array('Helper', 'AppHelper', 'ClassRegistry', 'Controller', 'Model'));
App::import('Helper', array('Recaptcha.Recaptcha', 'Html', 'Form'));
Mock::generatePartial('HtmlHelper', 'RecaptchaTestHtmlHelper', array('scriptBlock'));

/**
 * PostsTestController
 *
 * @package recaptcha
 * @subpackage recaptcha.tests.cases.helper
 */
class PostsTestController extends Controller {

/**
 * name property
 *
 * @var string 'Media'
 */
	public $name = 'PostsTest';

/**
 * uses property
 *
 * @var mixed null
 */
	public $uses = null;
}

Mock::generate('View', 'MockView');

/**
 * RecaptchaHelperTest
 *
 * @package recaptcha
 * @subpackage recaptcha.tests.cases.helpers
 */
class RecaptchaHelperTest extends CakeTestCase {

/**
 * setUp method
 *
 * @return void
 */
	function setup() {
		// "Global" keys for testing
		Configure::write('Recaptcha.publicKey', '6LdsGskSAAAAAPz645yuA9Bwj4BgBkWtWJZj_iWa');
		Configure::write('Recaptcha.privateKey', '6LdsGskSAAAAANEUP5yhcJ-XzbeUqSySGKfWH_Zm');

		

		$view = new View(new PostsTestController());
		$this->Recaptcha = new RecaptchaHelper();
		$this->Recaptcha->Html = new RecaptchaTestHtmlHelper();
		$this->Recaptcha->Form = new FormHelper();
		ClassRegistry::addObject('view', $view);
	}

/**
 * tearDown method
 *
 * @return void
 */
	function tearDown() {
		ClassRegistry::flush();
		unset($this->Recaptcha);
	}

/**
 * testSignupUrl
 *
 * @return void
 */
	function testSignupUrl() {
		$result = $this->Recaptcha->signupUrl('test');
		$expected = 'http://recaptcha.net/api/getkey?domain=' . WWW_ROOT . '&amp;app=test' ;
		$this->assertIdentical($expected, $result);
	}

/**
 * testDisplayDefault
 *
 * @return void
 */
	function testDisplayDefault() {
		$this->Recaptcha->Html->expectOnce('scriptBlock', array(
				'var RecaptchaOptions = {"theme":"red","lang":"en","custom_translations":[]}',
				array('inline' => false)));

		$result = $this->Recaptcha->display(array('publicKey' => Configure::read('Recaptcha.publicKey')));
		$expected = <<<TEXT
<div class="recaptcha"><script type="text/javascript" src="https://www.google.com/recaptcha/api/challenge?k=6LdsGskSAAAAAPz645yuA9Bwj4BgBkWtWJZj_iWa"></script>
				<noscript>
					<iframe src="https://www.google.com/recaptcha/api/noscript?k=6LdsGskSAAAAAPz645yuA9Bwj4BgBkWtWJZj_iWa" height="300" width="500" frameborder="0"></iframe><br/>
					<textarea name="recaptcha_challenge_field" rows="3" cols="40"></textarea>
					<input type="hidden" name="recaptcha_response_field" value="manual_challenge"/>
				</noscript></div>
TEXT;
		$this->assertIdentical($expected, $result);
	}

/**
 * testDisplayTheme
 *
 * @return void
 */
	function testDisplayTheme() {
		$this->Recaptcha->Html->expectOnce('scriptBlock', array(
				'var RecaptchaOptions = {"theme":"blackglass","lang":"en","custom_translations":[]}',
				array('inline' => false)));

		$result = $this->Recaptcha->display(array(
			'publicKey' => Configure::read('Recaptcha.publicKey'),
			'recaptchaOptions' => array(
				'theme' => 'blackglass')));
		$expected = <<<TEXT
<div class="recaptcha"><script type="text/javascript" src="https://www.google.com/recaptcha/api/challenge?k=6LdsGskSAAAAAPz645yuA9Bwj4BgBkWtWJZj_iWa"></script>
				<noscript>
					<iframe src="https://www.google.com/recaptcha/api/noscript?k=6LdsGskSAAAAAPz645yuA9Bwj4BgBkWtWJZj_iWa" height="300" width="500" frameborder="0"></iframe><br/>
					<textarea name="recaptcha_challenge_field" rows="3" cols="40"></textarea>
					<input type="hidden" name="recaptcha_response_field" value="manual_challenge"/>
				</noscript></div>
TEXT;
		$this->assertIdentical($expected, $result);
	}

/**
 * testDisplayError
 *
 * @return void
 */
	function testDisplayError() {
		$this->Recaptcha->Html->expectOnce('scriptBlock', array(
				'var RecaptchaOptions = {"theme":"red","lang":"en","custom_translations":[]}',
				array('inline' => false)));

		$result = $this->Recaptcha->display(array(
			'publicKey' => Configure::read('Recaptcha.publicKey'),
			'error' => 'testError'));
		$expected = <<<TEXT
<div class="recaptcha"><script type="text/javascript" src="https://www.google.com/recaptcha/api/challenge?k=6LdsGskSAAAAAPz645yuA9Bwj4BgBkWtWJZj_iWa"></script>
				<noscript>
					<iframe src="https://www.google.com/recaptcha/api/noscript?k=6LdsGskSAAAAAPz645yuA9Bwj4BgBkWtWJZj_iWa" height="300" width="500" frameborder="0"></iframe><br/>
					<textarea name="recaptcha_challenge_field" rows="3" cols="40"></textarea>
					<input type="hidden" name="recaptcha_response_field" value="manual_challenge"/>
				</noscript></div>
TEXT;
		$this->assertIdentical($expected, $result);
	}
}


views/helpers/recaptcha.php
<?php
/**
 * Copyright 2009-2010, Cake Development Corporation (http://cakedc.com)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright Copyright 2009-2010, Cake Development Corporation (http://cakedc.com)
 * @license MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

/**
 * CakePHP Recaptcha helper
 *
 * @package recaptcha
 * @subpackage recaptcha.views.helpers
 */
class RecaptchaHelper extends AppHelper {

/**
 * Secure API Url
 *
 * @var string
 */
	public $secureApiUrl = 'https://www.google.com/recaptcha/api';

/**
 * API Url
 *
 * @var string
 */
	public $apiUrl = 'http://www.google.com/recaptcha/api';

/**
 * View helpers
 *
 * @var array
 */
	public $helpers = array('Form', 'Html');

/**
 * Displays the Recaptcha input
 * 
 * @param array $options An array of options
 * 
 * ### Options:
 *
 * - `recaptchaOptions` assoc array of options to pass into RecaptchaOptions var, like 'theme', 'lang'
 *    or 'custom_translations' to runtime configure the widget.
 * 
 * @return string The resulting mark up
 * @access public
 */
	public function display($options = array()) {
		$defaults = array(
			'element' => null, 
			'publicKey' => Configure::read('Recaptcha.publicKey'),
			'error' => null,
			'ssl' => true,
			'error' => false,
			'div' => array(
				'class' => 'recaptcha'
			),
			'recaptchaOptions' => array(
				'theme' => 'red',
				'lang' => 'en',
				'custom_translations' => array(),
                                'callback' => 'Recaptcha.focus_response_field'
			)
		);
		$options = Set::merge($defaults, $options);
		extract($options);

		if ($ssl) {
			$server = $this->secureApiUrl;
		} else {
			$server = $this->apiUrl;
		}

		$errorpart = "";
		if ($error) {
			$errorpart = "&amp;error=" . $error;
		}

		if (!empty($element)) {
			$elementOptions = array();
			if (is_array($element)) {
				$keys = array_keys($element);
				$elementOptions = $element[$keys[0]];
			}
			$View = $this->__view();
			return $View->element($element, $elementOptions);
		}

		$jsonOptions = preg_replace('/"callback":"([^"\r\n]*)"/','"callback":$1',json_encode($recaptchaOptions));
		unset($recaptchaOptions);

		if (empty($this->params['isAjax'])) {
			$configScript = sprintf('var RecaptchaOptions = %s', $jsonOptions);
			$this->Html->scriptBlock($configScript, array('inline' => false));

			$script = '<script type="text/javascript" src="'. $server . '/challenge?k=' . $publicKey . '"></script>
				<noscript>
					<iframe src="'. $server . '/noscript?k=' . $publicKey . '" height="300" width="500" frameborder="0"></iframe><br/>
					<textarea name="recaptcha_challenge_field" rows="3" cols="40"></textarea>
					<input type="hidden" name="recaptcha_response_field" value="manual_challenge"/>
				</noscript>';

			if (!empty($error)) {
				$script .= $this->Form->error($error);
			}

			if ($options['div'] != false) {
				$script = $this->Html->tag('div', $script, $options['div']);
			}

			return $script;
		}

		$id = uniqid('recaptcha-');
		
		return	'<div id="'.$id.'"></div>' .
				'<script>
					if (window.Recaptcha == undefined) {
						(function() {
							var headID = document.getElementsByTagName("head")[0];
							var newScript = document.createElement("script");
							newScript.type = "text/javascript";
							newScript.onload = function() {
								Recaptcha.create("' . $publicKey . '", "' . $id . '", ' . $jsonOptions . ');
							};
							newScript.src = "'. $server . '/js/recaptcha_ajax.js"
							headID.appendChild(newScript);
						})();
					} else {
						setTimeout(\'Recaptcha.create("'.$publicKey.'", "'.$id.'", ' . $jsonOptions . ')\', 1000);
					}
				</script>';
	}

/**
 * Recaptcha signup URL
 *
 * @return string
 */
	public function signupUrl($appName = null) {
		return "http://recaptcha.net/api/getkey?domain=" . WWW_ROOT . '&amp;app=' . urlencode($appName);
	}

/**
 * AES Pad
 *
 * @param string value
 * @return string
 */
	private function __AesPad($val) {
		$blockSize = 16;
		$numpad = $blockSize - (strlen($val) % $blockSize);
		return str_pad($val, strlen ($val) + $numpad, chr($numpad));
	}

/**
 * AES Encryption
 *
 * @return string
 */
	private function __AesEncrypt($value, $key) {
		if (!function_exists('mcrypt_encrypt')) {
			throw new Exception(__d('recaptcha', 'To use reCAPTCHA Mailhide, you need to have the mcrypt php module installed.', true));
		}

		$mode = MCRYPT_MODE_CBC;
		$encryption = MCRYPT_RIJNDAEL_128;
		$value = $this->__AesPad($value);

		return mcrypt_encrypt($encryption, $key, $value, $mode, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0");
	}

/**
 * Mail-hide URL
 *
 * @return string base 64 encrypted string
 */
	private function __mailhideUrlbase64 ($x) {
		return strtr(base64_encode ($x), '+/', '-_');
	}

/**
 * Gets the reCAPTCHA Mailhide url for a given email
 *
 * @param $email
 * @return string
 */
	public function mailHideUrl($email = null) {
		$publicKey = Configure::read('Recaptcha.mailHide.publicKey');
		$privateKey = Configure::read('Recaptcha.mailHide.privateKey');

		if ($publicKey == '' || $publicKey == null || $privateKey == "" || $privateKey == null) {
			throw new Exception(__d('recaptcha', "You need to set a private and public mail hide key. Please visit http://mailhide.recaptcha.net/apikey", true));
		}

		$key = pack('H*', $privateKey);
		$cryptmail = $this->__AesEncrypt ($email, $key);

		return "http://mailhide.recaptcha.net/d?k=" . $publicKey . "&c=" . $this->__mailhideUrlbase64($cryptmail);
	}

/**
 * Get a part of the email to show
 *
 * Given johndoe@example,com return ["john", "example.com"].
 * the email is then displayed as john...@example.com
 *
 * @param string email
 * @return array
 */
	private function __hideEmailParts($email) {
		$array = preg_split("/@/", $email );

		if (strlen ($array[0]) <= 4) {
			$array[0] = substr ($array[0], 0, 1);
		} else if (strlen ($array[0]) <= 6) {
			$array[0] = substr ($array[0], 0, 3);
		} else {
			$array[0] = substr ($array[0], 0, 4);
		}
		return $array;
	}

/**
 * Gets html to display an email address given a public an private key to get a key go to:
 * http://mailhide.recaptcha.net/apikey
 *
 * @param string Email 
 * @return string
 */
	public function mailHide($email) {
		$emailparts = __hideEmailParts ($email);
		$url = $this->mailHideUrl($email);

		return htmlentities($emailparts[0]) . "<a href='" . htmlentities ($url) .
			"' onclick=\"window.open('" . htmlentities ($url) . "', '', 'toolbar=0,scrollbars=0,location=0,statusbar=0,menubar=0,resizable=0,width=500,height=300'); return false;\" title=\"Reveal this e-mail address\">...</a>@" . htmlentities ($emailparts [1]);
	}

/**
 * Get current view class
 *
 * @return object, View class
 */
	private function __view() {
		if (!empty($this->globalParams['viewInstance'])) {
			return $this->globalParams['viewInstance'];
		}
		return ClassRegistry::getObject('view');
	}

}


Model/Behavior/RecaptchaBehavior.php
<?php
/**
 * Copyright 2009-2014, Cake Development Corporation (http://cakedc.com)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright Copyright 2009-2014, Cake Development Corporation (http://cakedc.com)
 * @license MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

/**
 * CakePHP Recaptcha Behavior
 *
 * @package recaptcha
 * @subpackage recaptcha.models.behaviors
 */
class RecaptchaBehavior extends ModelBehavior {

/**
 * Settings array
 *
 * @var array
 */
	public $settings = array();

/**
 * Default settings
 *
 * @var array
 */
	public $defaults = array(
		'errorField' => 'recaptcha'
	);

/**
 * Setup
 *
 * @param Model $Model Model using this behavior
 * @param array $settings Configuration settings for $model
 * @return void
 */
	public function setup(Model $Model, $settings = array()) {
		if (!isset($this->settings[$Model->alias])) {
			$this->settings[$Model->alias] = $this->defaults;
		}
		$this->settings[$Model->alias] = array_merge($this->settings[$Model->alias], (is_array($settings) ? $settings : array()));
	}

/**
 * Validates the captcha responses status set by the component to the model
 *
 * @param Model $Model Model using this behavior
 * @return boolean
 * @see RecaptchaComponent::initialize()
 */
	public function validateCaptcha(Model $Model) {
		if (isset($Model->recaptcha) && $Model->recaptcha === false) {
			$Model->invalidate($this->settings[$Model->alias]['errorField'], $Model->recaptchaError);
		}
		return true;
	}

/**
 * Validates the captcha
 *
 * @param Model $Model Model using this behavior
 * @param array $options Options passed from Model::save()
 * @return void
 */
	public function beforeValidate(Model $Model, $options = array()) {
		$this->validateCaptcha($Model);
		return true;
	}

}

Controller/Component/RecaptchaComponent.php
<?php
/**
 * Copyright 2009-2014, Cake Development Corporation (http://cakedc.com)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright Copyright 2009-2014, Cake Development Corporation (http://cakedc.com)
 * @license MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

App::uses('HttpSocket', 'Network/Http');

/**
 * CakePHP Recaptcha component
 *
 * @package recaptcha
 * @subpackage recaptcha.controllers.components
 */

class RecaptchaComponent extends Component {

/**
 * Name
 *
 * @var string
 */
	public $Controller = null;

/**
 * Recaptcha API Url
 *
 * @var string
 */
	public $apiUrl = 'http://www.google.com/recaptcha/api/verify';

/**
 * Private API Key
 *
 * @var string
 */
	public $privateKey = '';

/**
 * Error coming back from Recaptcha
 *
 * @var string
 */
	public $error = null;

/**
 * Actions that should automatically checked for a recaptcha input
 *
 * @var array
 */
	public $actions = array();

/**
 * Settings
 *
 * @var array
 */
	public $settings = array();

/**
 * Default Options
 *
 * @var array
 */
	protected $_defaults = array(
		'errorField' => 'recaptcha',
		'actions' => array()
	);

 /**
 * Constructor
 *
 * @param ComponentCollection $collection A ComponentCollection this component can use to lazy load its components
 * @param array $settings Array of configuration settings
 * @return RecaptchaComponent
 */
	public function __construct(ComponentCollection $collection, $settings = array()) {
		parent::__construct($collection, $settings);
		$this->Controller = $collection->getController();
		$this->_defaults['modelClass'] = $this->Controller->modelClass;
		$this->settings = array_merge($this->_defaults, $settings);
		$this->actions = array_merge($this->actions, $this->settings['actions']);
		unset($this->settings['actions']);
	}

/**
 * Callback
 *
 * @param Controller $controller Controller with components to initialize
 * @param array $settings Array of configuration settings
 * @throws Exception Throws an exception if Recaptchas config is not present
 * @return void
 */
	public function initialize(Controller $controller, $settings = array()) {
		if ($controller->name === 'CakeError') {
			return;
		}
		$this->privateKey = Configure::read('Recaptcha.privateKey');
		$this->Controller = $controller;

		if (!isset($this->Controller->helpers['Recaptcha.Recaptcha'])) {
			$this->Controller->helpers[] = 'Recaptcha.Recaptcha';
		}

		if (empty($this->privateKey)) {
			throw new Exception(__d('recaptcha', "You must set your private Recaptcha key using Configure::write('Recaptcha.privateKey', 'your-key');!", true));
		}
	}

/**
 * Callback
 *
 * @param Controller $controller Controller with components to initialize
 * @return void
 */
	public function startup(Controller $controller) {
		extract($this->settings);
		$this->Controller->helpers[] = 'Recaptcha.Recaptcha';
		$this->Controller->{$modelClass}->Behaviors->load('Recaptcha.Recaptcha', array(
			'field' => $errorField
		));

		$this->Controller->{$modelClass}->recaptcha = true;
		if (in_array($this->Controller->action, $this->actions)) {
			if (!$this->verify()) {
				$this->Controller->{$modelClass}->recaptcha = false;
				$this->Controller->{$modelClass}->recaptchaError = $this->error;
			}
		}
	}

/**
 * Verifies the recaptcha input
 *
 * Please note that you still have to pass the result to the model and do
 * the validation there to make sure the data is not saved!
 *
 * @return boolean True if the response was correct
 */
	public function verify() {
		if (isset($this->Controller->request->data['recaptcha_challenge_field']) &&
			isset($this->Controller->request->data['recaptcha_response_field'])) {

			$response = $this->_getApiResponse();
			$response = explode("\n", $response->body());

			if (empty($response[0])) {
				$this->error = __d('recaptcha', 'Invalid API response, please contact the site admin.', true);
				return false;
			}

			if ($response[0] === 'true') {
				return true;
			}

			if ($response[1] === 'incorrect-captcha-sol') {
				$this->error = __d('recaptcha', 'Incorrect captcha', true);
			} else {
				$this->error = $response[1];
			}
		}
		return false;
	}

/**
 * Queries the Recaptcha API and and returns the raw response
 *
 * @return string
 */
	protected function _getApiResponse() {
		$Socket = new HttpSocket();
		return $Socket->post($this->apiUrl, array(
			'privatekey' => $this->privateKey,
			'remoteip' => env('REMOTE_ADDR'),
			'challenge' => $this->Controller->request->data['recaptcha_challenge_field'],
			'response' => $this->Controller->request->data['recaptcha_response_field']
		));
	}

}


View/Helper/RecaptchaHelper.php
<?php
/**
 * Copyright 2009-2014, Cake Development Corporation (http://cakedc.com)
 *
 * Licensed under The MIT License
 * Redistributions of files must retain the above copyright notice.
 *
 * @copyright Copyright 2009-2014, Cake Development Corporation (http://cakedc.com)
 * @license MIT License (http://www.opensource.org/licenses/mit-license.php)
 */

/**
 * CakePHP Recaptcha helper
 *
 * @package recaptcha
 * @subpackage recaptcha.views.helpers
 */
class RecaptchaHelper extends AppHelper {

/**
 * Secure API Url
 *
 * @var string
 */
	public $secureApiUrl = 'https://www.google.com/recaptcha/api';

/**
 * API Url
 *
 * @var string
 */
	public $apiUrl = 'http://www.google.com/recaptcha/api';

/**
 * View helpers
 *
 * @var array
 */
	public $helpers = array('Form', 'Html');

/**
 * Displays the Recaptcha input
 *
 * @param array $options An array of options
 *
 * ### Options:
 *
 * - `element` String, name of the view element that can be used instead of the hardcoded HTML structure from this helper
 * - `publicKey` String, default is read from Configure::read('Recaptcha.publicKey'), you can override it here
 * - `error` String, optional error message that is displayed using Form::error()
 * - `ssl` Boolean, use SSL or not, default is true
 * - `div` Array of options for the div tag the recaptcha is wrapped with, set to false if you want to disable it
 * - `recaptchaOptions` assoc array of options to pass into RecaptchaOptions var, like 'theme', 'lang'
 *    or 'custom_translations' to runtime configure the widget.
 *
 * @return string The resulting mark up
 * @access public
 */
	public function display($options = array()) {
		$defaults = array(
			'element' => null,
			'publicKey' => Configure::read('Recaptcha.publicKey'),
			'error' => null,
			'ssl' => true,
			'error' => false,
			'div' => array(
				'class' => 'recaptcha'),
				'recaptchaOptions' => array(
					'theme' => 'red',
					'lang' => 'en',
					'custom_translations' => array()
				)
		);

		$options = Set::merge($defaults, $options);
		extract($options);

		if ($ssl) {
			$server = $this->secureApiUrl;
		} else {
			$server = $this->apiUrl;
		}

		$errorpart = "";
		if ($error) {
			$errorpart = "&amp;error=" . $error;
		}

		if (!empty($element)) {
			$elementOptions = array();
			if (is_array($element)) {
				$keys = array_keys($element);
				$elementOptions = $element[$keys[0]];
			}

			return $this->View->element($element, $elementOptions);
		}

		$jsonOptions = preg_replace('/"callback":"([^"\r\n]*)"/', '"callback":$1', json_encode($recaptchaOptions));
		unset($recaptchaOptions);

		if (empty($this->params['isAjax'])) {
			$configScript = sprintf('var RecaptchaOptions = %s', $jsonOptions);
			echo $this->Html->scriptBlock($configScript);

			$script = '';
			$script .= '<script type="text/javascript" src="' . $server . '/challenge?k=' . $publicKey . '"></script>';
			$script .= '<noscript>';
			$script .= '	<iframe src="' . $server . '/noscript?k=' . $publicKey . '" height="300" width="500" frameborder="0"></iframe><br/>';
			$script .= '	<textarea name="recaptcha_challenge_field" rows="3" cols="40"></textarea>';
			$script .= '  <input type="hidden" name="recaptcha_response_field" value="manual_challenge"/>';
			$script .= '</noscript>';

			if (!empty($error)) {
				$script .= $this->Form->error($error);
			}

			if ($options['div'] != false) {
				$script = $this->Html->tag('div', $script, $options['div']);
			}

			$this->Form->unlockField('recaptcha_challenge_field');
			$this->Form->unlockField('recaptcha_response_field');

			return $script;
		}

		$id = uniqid('recaptcha-');

		$script = '';
		$script .= '<div id="' . $id . '"></div>';
		$script .= '<script>';
		$script .= 'if (window.Recaptcha == undefined) {';
		$script .= '  (function() {';
		$script .= '	  var headID = document.getElementsByTagName("head")[0];';
		$script .= '		var newScript = document.createElement("script");';
		$script .= '		newScript.type = "text/javascript";';
		$script .= '		newScript.onload = function() {';
		$script .= '			Recaptcha.create("' . $publicKey . '", "' . $id . '", ' . $jsonOptions . ');';
		$script .= '		  Recaptcha.focus_response_field();';
		$script .= '		};';
		$script .= '		newScript.src = "' . $server . '/js/recaptcha_ajax.js"';
		$script .= '	  headID.appendChild(newScript);';
		$script .= '  })();';
		$script .= '} else {';
		$script .= '  setTimeout(\'Recaptcha.create("' . $publicKey . '", "' . $id . '", ' . $jsonOptions . ')\', 1000);';
		$script .= '}';
		$script .= '</script>';

		return $script;
	}

/**
 * Recaptcha signup URL
 *
 * @param string $appName An application name
 * @return string A signup url
 */
	public function signupUrl($appName = null) {
		return "http://recaptcha.net/api/getkey?domain=" . WWW_ROOT . '&amp;app=' . urlencode($appName);
	}

/**
 * AES Pad
 *
 * @param string $val A value to pad
 * @return string
 */
	private function __aesPad($val) {
		$blockSize = 16;
		$numpad = $blockSize - (strlen($val) % $blockSize);
		return str_pad($val, strlen($val) + $numpad, chr($numpad));
	}

/**
 * AES Encryption
 *
 * @param string $value A value
 * @param string $key A key to use
 * @return string
 * @throws Exception
 */
	private function __aesEncrypt($value, $key) {
		if (!function_exists('mcrypt_encrypt')) {
			throw new Exception(__d('recaptcha', 'To use reCAPTCHA Mailhide, you need to have the mcrypt php module installed.', true));
		}

		$mode = MCRYPT_MODE_CBC;
		$encryption = MCRYPT_RIJNDAEL_128;
		$value = $this->__aesPad($value);

		return mcrypt_encrypt($encryption, $key, $value, $mode, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0");
	}

/**
 * Mail-hide URL
 *
 * @param string $x An input string
 * @return string A base 64 encrypted string
 */
	private function __mailhideUrlbase64($x) {
		return strtr(base64_encode($x), '+/', '-_');
	}

/**
 * Gets the reCAPTCHA Mailhide url for a given email
 *
 * @param string $email An email address
 * @return string
 * @throws Exception
 */
	public function mailHideUrl($email = null) {
		$publicKey = Configure::read('Recaptcha.mailHide.publicKey');
		$privateKey = Configure::read('Recaptcha.mailHide.privateKey');

		if ($publicKey == '' || $publicKey == null || $privateKey == "" || $privateKey == null) {
			throw new Exception(__d('recaptcha', "You need to set a private and public mail hide key. Please visit http://mailhide.recaptcha.net/apikey", true));
		}

		$key = pack('H*', $privateKey);
		$cryptmail = $this->__aesEncrypt($email, $key);

		return "http://mailhide.recaptcha.net/d?k=" . $publicKey . "&c=" . $this->__mailhideUrlbase64($cryptmail);
	}

/**
 * Get a part of the email to show
 *
 * Given johndoe@example,com return ["john", "example.com"].
 * the email is then displayed as john...@example.com
 *
 * @param string $email an email address
 * @return array
 */
	private function __hideEmailParts($email) {
		$array = preg_split("/@/", $email );

		if (strlen($array[0]) <= 4) {
			$array[0] = substr($array[0], 0, 1);
		} elseif (strlen($array[0]) <= 6) {
			$array[0] = substr($array[0], 0, 3);
		} else {
			$array[0] = substr($array[0], 0, 4);
		}
		return $array;
	}

/**
 * Gets html to display an email address given a public an private key to get a key go to:
 * http://mailhide.recaptcha.net/apikey
 *
 * @param string $email An email address
 * @return string
 */
	public function mailHide($email) {
		$emailparts = $this->__hideEmailParts($email);
		$url = $this->mailHideUrl($email);

		return htmlentities($emailparts[0]) . "<a href='" . htmlentities($url) .
			"' onclick=\"window.open('" . htmlentities($url) . "', '', 'toolbar=0,scrollbars=0,location=0,statusbar=0,menubar=0,resizable=0,width=500,height=300'); return false;\" title=\"Reveal this e-mail address\">...</a>@" . htmlentities($emailparts[1]);
	}

}
